draft-pauly-taps-arch.md

```diff
@@ -237,21 +237,19 @@ Beyond the basic objects, there are several high-level groups of actions that an
+* Preconnection: A Preconnection object is a representation of a potential connection. It has state that describes parameters of a Connection that might exist in the future: the LocalEndpoint from which that Connection will be established, the RemoteEndpoint to which it will connect, and Path Selection Properties, Protocol Selection Properties, and Specific Protocol Properties that influence the choice of transport that Connection will use. A Preconnection can be fully specified and represent a single possible Connection, or it can be partially specified such that it represents a family of possible Connections. The LocalEndpoint MUST be specified if the Preconnection is used to Listen() for incoming connections, but is OPTIONAL if it is used to Initiate() connections. The RemoteEndpoint MUST be specified in the Preconnection is used to Initiate() connections, but is OPTIONAL if it is used to Listen() for incoming connections.
 
-* Connection: A Connection object represents an active transport protocol instance that can send and/or receive Content between a Local Endpoint and a Remote Endpoint. It holds state pertaining to the underlying transport protocol instance and any ongoing data transfer. This represents, for example, an active connection in a connection-oriented protocol such as TCP, or a fully-specified 5-tuple for a connectionless protocol such as UDP.
+* Connection: A Connection object represents an active transport protocol instance that can send and/or receive Content between a LocalEndpoint and a RemoteEndpoint. It holds state pertaining to the underlying transport protocol instance and any ongoing data transfer. This represents, for example, an active connection in a connection-oriented protocol such as TCP, or a fully-specified 5-tuple for a connectionless protocol such as UDP.
 
-* Listener: A Listener object accepts incoming transport protocol connections from Remote Endpoints  and generates corresponding Connection objects. It is created from a Preconnection object that specifies the type of incoming connections it will accept.
+* Listener: A Listener object accepts incoming transport protocol connections from RemoteEndpoints  and generates corresponding Connection objects. It is created from a Preconnection object that specifies the type of incoming connections it will accept.
 
 ### Pre-Establishment {#preestablishment}
 
-* Endpoint: An Endpoint represents one side of a transport connection. It is a concept that represents how an application views itself or a peer that may vary in levels of specificity. Examples include "IP address + port", "hostname + port", and "DNS service name".
+* RemoteEndpoint: The RemoteEndpoint represents the application's name for a peer that can participate in a transport connection. For example, a DNS name for the peer.
```

Comment:
I don't think a DNS name is sufficient for an Endpoint, unless that's referring to the DNS Service name. It needs to be a DNS hostname + port. I think listing out the types as we had with endpoint is still necessary and useful.

My concern with removing the overall definition of an Endpoint is that we lose the common place to discuss the fact that an endpoint is not simply one type (hostname + port, or address + port, or anything else), but an abstract representation of something that may correspond via resolution or advertisement to one or more other endpoints.

What precisely is the advantage of removing the definition of the concept of an endpoint? It's still a term that is not really standardized, and without defining it, I think we're losing a fundamental type. It's essentially the equivalent of sockaddr for sockets—a type that indeed has meaning in multiple contexts, not just in remote/local dichotomies. If we don't specify this carefully, we risk interpretations of endpoints to be just one set of things (like how sockets used just addresses) that will pigeonhole us later on.

Reply:
on reflection -- i think one of the reasons I thought of this as an implementation detail is precisely because it's a `sockaddr`, but this...

>> What precisely is the advantage of removing the definition of the concept of an endpoint? It's still a term that is not really standardized, and without defining it, I think we're losing a fundamental type.

... is compelling to me.

Reply:
Right, just as we want to live a post-sockets world (lower case post-sockets, here), we also want to live in a post-sockaddr world. 

For example, dealing with byte structures like sockaddrs in languages like Swift is really a pain. Using an Endpoint object with methods that replace the basic sockaddr fields is way easier.

Reply:
Remote Endpoint should be defined by name and port, I agree. That was an oversight that I've fixed. 

For Endpoint, we need to be clear that RemoteEndpoint is an abstract type that can represent several different types of remote endpoint, and that name+port is just one example. Similarly for LocalEndpoint. And I'm certainly not advocating `struct sockaddr` style anything. But, it's still not clear to me what a generic Endpoint would include? 

I expect `LocalEndpoint` and `RemoteEndpoint` may well have some common methods (e.g., they'd both implement `Display` in a Rust implementation), but I don't see use cases for a generic `Endpoint`. All the cases that I can immediately think of, it matters whether it's a local or remote endpoint, so having a common super type doesn't give us anything.

This is not to say that I'm against `Endpoint` in principle, I'd just like a use case before adding it.

Reply:
I agree that a raw Endpoint that isn't Remote or Local may not be used in any of the externally-facing APIs to create a Connection. From an API standpoint, it's fine to require a specific type of the general class.

All I'm arguing is that in our list of concepts and definitions, we need to define that there is a basic concept of an Endpoint, that isn't necessarily local or remote, since local and remote only become relevant in the context of a connection, or some communication. Having a definition of Remote and Local Endpoints without ever saying what an endpoint is (or duplicating the definitions, saying Local: See the generic parts of Remote) seems unnecessary.

Also, from an implementation standpoint, the Endpoint really isn't as trivial as you're saying. It's not just a matter of displaying. Since we don't use sockaddrs in the exposure of the address form of endpoint, the Endpoint does the following:
- Can create hostname/address from a string, which automatically parses into a hostname, IPv4, or IPv6 address. If the IPv6 address is a mapped v4 address, it parses that out.
- Create port with a string or number ("http" or "80")
- Extract address byte encoding for use in forming packets
- Extract port in either network or host byte order
- Extract IP address family
- Do subnet or DNS name segment matching and comparisons between endpoints (are these two in the same slash #)
- Allow definition of custom endpoint subclasses that can resolve into standard types

All of these operations are useful for both local and remote endpoints, such as when you are filling out or parsing headers within the protocol stack. If as an implementation, I was told to make the LocalEndpoint and RemoteEndpoint completely different classes in a strongly typed language, I would need to redefine all of these? If they're not subclasses of the same Endpoint generic class, at least they both implement the template/protocol of an Endpoint so we can re-use some logic.

To that end, defining the concept of Endpoint first, then explaining Remote and Local Endpoint, still makes sense in the architecture.